pytorch_lightning/trainer/states.py

```diff
@@ -11,50 +11,51 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+from dataclasses import dataclass
+from typing import Optional
 
 from pytorch_lightning.utilities import LightningEnum
 
 
-class TrainerState(LightningEnum):
-    """ State for the :class:`~pytorch_lightning.trainer.trainer.Trainer`
-    to indicate what is currently or was executed. It follows the user-called
-    functions such as `trainer.fit()` and `trainer.test().
+class TrainerStatus(LightningEnum):
+    """Status of the :class:`~pytorch_lightning.trainer.trainer.Trainer`"""
+    INITIALIZING = 'initializing'  # trainer creation
+    RUNNING = 'running'
+    FINISHED = 'finished'
+    INTERRUPTED = 'interrupted'
+
+    @property
+    def stopped(self) -> bool:
+        return self in (self.FINISHED, self.INTERRUPTED)
+
 
-    >>> # you can compare the type with a string
-    >>> TrainerState.FITTING == 'fit'
-    True
-    >>> # which is case insensitive
-    >>> TrainerState.FINISHED == 'FINISHED'
-    True
+class TrainerFn(LightningEnum):
+    """
+    Functions of the :class:`~pytorch_lightning.trainer.trainer.Trainer`
+    functions such as ``trainer.fit()`` and ``trainer.test()``.
     """
-    INITIALIZING = 'initializing'  # trainer creation
     FITTING = 'fit'  # trainer.fit()
     VALIDATING = 'validate'  # trainer.validate()
     TESTING = 'test'  # trainer.test()
     PREDICTING = 'predict'  # trainer.predict()
     TUNING = 'tune'  # trainer.tune()
-    FINISHED = 'finished'
-    INTERRUPTED = 'interrupted'
 
     @property
-    def stopped(self) -> bool:
-        return self in (self.FINISHED, self.INTERRUPTED)
+    def _setup_fn(self) -> 'TrainerFn':
+        """
+        'fit' is used instead of 'tune' as there aren't 'tune_dataloaders'.
 
-    @property
-    def running(self) -> bool:
-        return self in (self.FITTING, self.VALIDATING, self.TESTING, self.PREDICTING, self.TUNING)
+        This is used for the ``setup()`` and ``teardown()`` hooks
+        """
+        return TrainerFn.FITTING if self == TrainerFn.TUNING else self
 
```

Comment:
the tune/fit relationship is the only aspect of this PR that's odd to me, but that's beyond the scope of this PR

Reply:
It is, also very error prone.

Tuning could use a refactor in general.